cmd/cmd_utils.go

```diff
@@ -0,0 +1,200 @@
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	e "github.com/cloudposse/atmos/internal/exec"
+	c "github.com/cloudposse/atmos/pkg/config"
+	u "github.com/cloudposse/atmos/pkg/utils"
+	"github.com/spf13/cobra"
+	"os"
+	"strings"
+	"text/template"
+)
+
+var (
+	// This map contains the existing atmos top-level commands
+	// All custom top-level commands will be checked against this map in order to not override `atmos` top-level commands,
+	// but just add subcommands to them
+	existingTopLevelCommands = map[string]*cobra.Command{
+		"terraform": terraformCmd,
+		"helmfile":  helmfileCmd,
+		"describe":  describeCmd,
+		"aws":       awsCmd,
+		"validate":  validateCmd,
+		"vendor":    vendorCmd,
+		"workflow":  workflowCmd,
+		"version":   versionCmd,
+	}
+)
+
+func processCustomCommands(commands []c.Command, parentCommand *cobra.Command, topLevel bool) error {
+	var command *cobra.Command
+
+	for _, commandConfig := range commands {
+		if _, exist := existingTopLevelCommands[commandConfig.Name]; exist && topLevel {
+			command = existingTopLevelCommands[commandConfig.Name]
+		} else {
+			// Deep-copy the slices because of the automatic closure in the `Run` function of the command.
+			// https://www.calhoun.io/gotchas-and-common-mistakes-with-closures-in-go/
+			// It will make a closure on the new local variables which are different in each iteration.
+			customCommandSteps := make([]string, len(commandConfig.Steps))
+			copy(customCommandSteps, commandConfig.Steps)
+			customCommandArguments := make([]c.CommandArgument, len(commandConfig.Arguments))
+			copy(customCommandArguments, commandConfig.Arguments)
+			customCommandFlags := make([]c.CommandFlag, len(commandConfig.Flags))
+			copy(customCommandFlags, commandConfig.Flags)
+			customEnvVars := make([]c.CommandEnv, len(commandConfig.Env))
+			copy(customEnvVars, commandConfig.Env)
+
+			var customCommand = &cobra.Command{
+				Use:   commandConfig.Name,
+				Short: commandConfig.Description,
+				Long:  commandConfig.Description,
+				Run: func(cmd *cobra.Command, args []string) {
+					var err error
+					var t *template.Template
+
+					if len(args) != len(customCommandArguments) {
+						err := fmt.Errorf("invalid number of arguments, %d argement(s) required", len(customCommandArguments))
+						u.PrintErrorToStdErrorAndExit(err)
+					}
+
+					// Execute command's customCommandSteps
+					for i, step := range customCommandSteps {
+						// Prepare template data for arguments
+						argumentsData := map[string]string{}
+						for ix, arg := range customCommandArguments {
+							argumentsData[arg.Name] = args[ix]
+						}
+
+						// Prepare template data for flags
+						flags := cmd.Flags()
+						flagsData := map[string]string{}
+						for _, fl := range customCommandFlags {
+							if fl.Type == "" || fl.Type == "string" {
```

Comment:
What happens if the type is not string ?

Reply:
we support strings only for now.
`string` handle everything with key and val.
`bool` is just a flag name/key w/o a value, we'll need to address that special case.
That can be improved in follow up PRs.